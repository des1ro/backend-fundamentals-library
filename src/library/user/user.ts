import { randomUUID } from "crypto";

export class User {
  private readonly uuid: string = randomUUID();
  private credit: number = 0;
  constructor(private readonly name: string) {}
  getName(): string {
    return this.name;
  }
  getUuid(): string {
    return this.uuid;
  }
  resetCredit() {
    this.credit = 0; // Czy tu powinienem wrzucić credit do construktora, (private readonly credit) i chcąc zmiany tworzyć new User?
  }
  getCredit() {
    return this.credit;
  }
  setCredit(credit: number): void {
    this.credit = credit;
  }
}
