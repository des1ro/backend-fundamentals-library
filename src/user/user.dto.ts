import { randomUUID } from "crypto";

export class User {
  readonly uuid: string = randomUUID();
  constructor(
    readonly name: string,
    private dueDate = 0,
    private penaltyPoints = 0
  ) {}
  getDueDate(): number {
    return this.dueDate;
  }
  getPenaltyPoints(): number {
    return this.penaltyPoints;
  }
  setDueDate(dueDate: number): void {
    this.dueDate = dueDate;
  }
  setPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints = penaltyPoints;
  }
}
