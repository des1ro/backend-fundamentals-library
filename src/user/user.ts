import { randomUUID } from "crypto";
import { isAfter } from "date-fns";
export class User {
  readonly uuid: string = randomUUID();
  private dueDate?: number;
  private penaltyPoints = 0;
  constructor(readonly name: string) {}
  getPenaltyPoints() {
    return this.penaltyPoints;
  }
  addPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints += penaltyPoints;
  }
  subtractPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints -= penaltyPoints;
  }
  getDueDate() {
    return this.dueDate || 0;
  }
  setDueDate(date: number): void {
    this.dueDate = date;
  }
  isBanned(): boolean {
    return isAfter(this.getDueDate(), new Date());
  }
}
