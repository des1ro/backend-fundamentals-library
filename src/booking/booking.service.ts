import { isAfter, addMonths } from "date-fns";
import { Book } from "../book/book.dto";
import { LibraryError } from "../exceptions/library.exceptions";
import { User } from "../user/user.dto";

export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {}
  addBookToUser(book: Book): void {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    this.booksBorrowDate.delete(returnedBook);
  }
  getBooksDatesArray(): Date[] {
    return Array.from(this.booksBorrowDate.values());
  }
}
