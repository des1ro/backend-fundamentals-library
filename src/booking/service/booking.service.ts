import { addMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { LibraryError } from "../../exceptions/library.exceptions";
import { User } from "../../user/user";
export class Booking {
  private readonly penaltyPointsPerBook: number;
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {
    this.penaltyPointsPerBook = 10;
  }
  addBookToUser(book: Book): void {
    if (this.user.isBanned()) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    const bookBorrowDateLimit = addMonths(date, 1);
    const bookIsAfterDueDate = isAfter(new Date(), bookBorrowDateLimit);
    this.booksBorrowDate.delete(returnedBook);
    if (bookIsAfterDueDate) {
      this.user.subtractPenaltyPoints(this.penaltyPointsPerBook);
    }
  }
  has(book: Book): boolean {
    return this.booksBorrowDate.has(book);
  }
  _getBooksDatesArray(): Date[] {
    const datesArray = Array.from(this.booksBorrowDate.values());
    return datesArray;
  }
}
