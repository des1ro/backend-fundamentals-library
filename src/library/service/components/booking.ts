import { add, addMonths, differenceInMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
// sprawdzać przy oddawaniu czy książka po terminie
// wszystko w bookingu
// user ma mieć
export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {}
  getUser(): User {
    return this.user;
  }
  isUserBanned(
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): boolean {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      return true;
    }
    const userPenaltyPoints = this.checkUserPenaltyPoints(
      bookBorrowMonthsLimit,
      penaltyPointsLimit
    );
    if (userPenaltyPoints > 0) {
      const banMonths = userPenaltyPoints / 10;
      this.user.setDueDate(addMonths(new Date(), banMonths));
      return true;
    }
    return false;
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned(bookBorrowMonthsLimit, penaltyPointsLimit)) {
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERRMISION_DENIED",
      message: "User is banned!",
    });
  }
  checkUserPenaltyPoints(
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): number {
    let penaltyPoints = 0;
    this.booksDueDate.forEach((bookDate) => {
      const difference = differenceInMonths(new Date(), bookDate);
      if (difference >= bookBorrowMonthsLimit)
        penaltyPoints += penaltyPointsPerBook;
    });
    this.user.addPenaltyPoints(penaltyPoints);
    return this.user.getPenaltyPoints();
  }

  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): void {
    const isUserBanned = this.isUserBanned(
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (this.booksDueDate.has(returnedBook)) {
      const date = this.booksDueDate.get(returnedBook);
      if (date && isAfter(new Date(), addMonths(date, 1)) && isUserBanned) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
      }
      this.booksDueDate.delete(returnedBook);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  getBooksDueDate(): Map<Book, Date> {
    return this.booksDueDate;
  }
  getUserDueDate() {
    return this.user.getDueDate();
  }
  getUserPenaltyPoints() {
    return this.user.getPenaltyPoints();
  }
}
