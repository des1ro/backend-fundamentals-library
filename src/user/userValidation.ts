import { isAfter } from "date-fns";
import { User } from "./user";

export class UserValidation {
  constructor(private readonly user: User) {}
  isUserBanned() {
    return isAfter(this.user.getDueDate(), new Date());
  }
}
