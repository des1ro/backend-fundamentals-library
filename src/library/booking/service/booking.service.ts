import { isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
import { PenaltyPoints } from "./components/PenaltyPoints";
export class Booking {
  private readonly penaltyPoints: PenaltyPoints;
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {
    this.penaltyPoints = new PenaltyPoints(this.user);
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned()) {
      const penaltyPoints = this.penaltyPoints.calculatePenaltyPoints(
        this.booksDueDate,
        bookBorrowMonthsLimit,
        penaltyPointsPerBook
      );
      if (penaltyPoints > penaltyPointsLimit) {
        this.penaltyPoints.updateUserPenaltyPointsAndBan(penaltyPoints);
        throw new LibraryError({
          name: "PERMISION_DENIED",
          message: "User is banned!",
        });
      }
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERMISION_DENIED",
      message: "User is banned!",
    });
  }
  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number,
    penaltyPointsLimit: number
  ): void {
    const penaltyPoints = this.penaltyPoints.calculatePenaltyPoints(
      this.booksDueDate,
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (penaltyPoints > penaltyPointsLimit) {
      this.penaltyPoints.updateUserPenaltyPointsAndBan(penaltyPoints);
    }
    if (this.booksDueDate.has(returnedBook)) {
      this.booksDueDate.delete(returnedBook);
      if (this.isUserBanned() && penaltyPoints > 0) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
        return;
      }
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  has(book: Book): boolean {
    return this.booksDueDate.has(book);
  }
  private isUserBanned(): boolean {
    return isAfter(this.user.getDueDate(), new Date());
  }
}
