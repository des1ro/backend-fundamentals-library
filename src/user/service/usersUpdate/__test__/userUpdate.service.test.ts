import { addDays, addMonths } from "date-fns";
import { Booking } from "../../../../booking/booking.service";
import { User } from "../../../user.dto";
import { UsersUpdate } from "../usersUpdate.service";
jest.mock("crypto", () => ({
  randomUUID: jest.fn().mockReturnValue("test-uuid"),
}));
describe("UserUpdate test suite", () => {
  let objectUnderTest: UsersUpdate;
  let mockedUserOne: User;
  let mockedUserTwo: User;
  let mockedBookingOne: Booking;
  let mockedBookingTwo: Booking;
  let mockedDatesBeforeLimit: Date[];
  let mockedDatesAfterLimit: Date[];
  let spyUser: jest.SpyInstance;
  let spyBooking: jest.SpyInstance;
  beforeEach(() => {
    objectUnderTest = new UsersUpdate();
  });
  beforeAll(() => {
    mockedUserOne = new User("Mocked user one");
    mockedUserTwo = new User("Mocked user two");
    mockedBookingOne = new Booking(mockedUserOne);
    mockedBookingTwo = new Booking(mockedUserTwo);
    mockedDatesBeforeLimit = [
      addDays(new Date(), 20),
      addDays(new Date(), 24),
      addDays(new Date(), 10),
    ];
    mockedDatesAfterLimit = [
      addDays(new Date(), 32),
      addDays(new Date(), 33),
      addMonths(new Date(), 1),
    ];
  });
  afterEach(() => {
    jest.useRealTimers();
    jest.resetAllMocks();
  });
  it("Should return and not add penaltyPoints user if user is banned", () => {
    //Given
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 0, 20, 12, 30, 0, 0));
    const testDate = addDays(new Date(), 1).getTime();
    spyUser = jest.spyOn(mockedUserOne, "getDueDate").mockReturnValue(testDate);
    spyBooking = jest
      .spyOn(mockedBookingOne, "getBooksDatesArray")
      .mockReturnValue(mockedDatesAfterLimit);
    const mockedMap = new Map<User, Booking>();
    //When
    mockedMap.set(mockedUserOne, mockedBookingOne);
    objectUnderTest.update(mockedMap);
    //Then
    expect(mockedUserOne).toMatchSnapshot();
    expect(spyUser).toBeCalledTimes(1);
    expect(spyBooking).toBeCalledTimes(0);
  });
  it("Should return and not add penaltyPoints if users books not reach limit", () => {
    //Given
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 0, 20, 12, 30, 0, 0));
    const testDate = new Date().getTime();
    spyUser = jest.spyOn(mockedUserOne, "getDueDate").mockReturnValue(testDate);
    spyBooking = jest
      .spyOn(mockedBookingOne, "getBooksDatesArray")
      .mockReturnValue(mockedDatesBeforeLimit);
    const mockedMap = new Map<User, Booking>();
    //When
    mockedMap.set(mockedUserOne, mockedBookingOne);
    objectUnderTest.update(mockedMap);
    //Then
    expect(mockedUserOne).toMatchSnapshot();
    expect(spyUser).toBeCalledTimes(1);
    expect(spyBooking).toBeCalledTimes(1);
  });
  it("Should add penaltyPoints and ban user if books reach limit", () => {
    //Given
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 0, 20, 12, 30, 0, 0));
    const testDate = new Date().getTime();
    spyUser = jest.spyOn(mockedUserOne, "getDueDate").mockReturnValue(testDate);
    spyBooking = jest
      .spyOn(mockedBookingOne, "getBooksDatesArray")
      .mockReturnValue(mockedDatesAfterLimit);
    const mockedMap = new Map<User, Booking>();
    //When
    mockedMap.set(mockedUserOne, mockedBookingOne);
    objectUnderTest.update(mockedMap);

    //Then
    expect(mockedUserOne).toMatchSnapshot();
    expect(spyBooking).toBeCalledTimes(1);
    expect(spyUser).toBeCalledTimes(1);
  });
});
