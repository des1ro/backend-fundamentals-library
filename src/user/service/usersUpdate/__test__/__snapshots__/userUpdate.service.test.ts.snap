// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserUpdate test suite Should add penaltyPoints and ban user if books reach limit 1`] = `
User {
  "dueDate": 1689849000000,
  "getDueDate": [MockFunction] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": 1674214200000,
      },
    ],
  },
  "name": "Mocked user one",
  "penaltyPoints": 60,
  "uuid": "test-uuid",
}
`;

exports[`UserUpdate test suite Should return and not add penaltyPoints if users books not reach limit 1`] = `
User {
  "dueDate": 1681986600000,
  "getDueDate": [MockFunction] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": 1674214200000,
      },
    ],
  },
  "name": "Mocked user one",
  "penaltyPoints": 30,
  "uuid": "test-uuid",
}
`;

exports[`UserUpdate test suite Should return and not add penaltyPoints user if user is banned 1`] = `
User {
  "dueDate": 0,
  "getDueDate": [MockFunction] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": 1674300600000,
      },
    ],
  },
  "name": "Mocked user one",
  "penaltyPoints": 0,
  "uuid": "test-uuid",
}
`;
